### ISSUE TITLE
SDPA miscompiles packed-row encoder attention at extents 128 and >256 (numerical regression from #4551)

### ISSUE THREAD

**Issue by gkumbhat** [torch-spyre/torch-spyre#4824]
`F.scaled_dot_product_attention` returns numerically wrong results for packed-row
encoder attention at several sequence extents. This is a **regression from #4551**
(`aabfe9cd`, "Rewrite SDPA tiling with nested `for_each_tile`"): its parent `ffe4d6ee`
is correct at every affected shape. It is present on current `main`, is silent (no
error, no warning, no fallback), and at `extent=128` it is **intermittent**.

Same commit as the ~2.9x slowdown in #4784. That was fixed by #4790; this was not.

## Repro

Pure torch-spyre, no vLLM. Fixed seed, zero mask, fp16.

```python
import sys
import torch
import torch.nn.functional as F
import torch_spyre  # noqa: F401

group, extent = int(sys.argv[1]), int(sys.argv[2])
H, D = 12, 64                      # granite-embedding-278m-multilingual
rows = group * extent
gen = torch.Generator().manual_seed(4676)
q, k, v = (torch.randn((rows, H, D), dtype=torch.float16, generator=gen) for _ in range(3))
mask = torch.zeros(group, 1, 1, extent, dtype=torch.float16)

def f(qr, kr, vr, m):
    up = lambda x: x.reshape(group, extent, H, D).transpose(1, 2)
    a = F.scaled_dot_product_attention(up(qr), up(kr), up(vr), attn_mask=m, scale=D ** -0.5)
    return a.transpose(1, 2).reshape(rows, H, D)

ref = f(q.float(), k.float(), v.float(), mask.float())          # fp32 CPU reference
dev = torch.compile(f, fullgraph=True, dynamic=False)(
    q.to("spyre"), k.to("spyre"), v.to("spyre"), mask.to("spyre")).cpu().float()
err = (dev - ref).abs().max().item()
print(f"RESULT group={group} extent={extent} max_abs_err={err:.5f} "
      f"{'WRONG' if err > 0.1 else 'ok'}")
```

```bash
rm -rf /tmp/torchinductor_*        # per invocation
python repro.py 4 512             # -> WRONG
python repro.py 4 128             # -> WRONG on most runs, ok on some
```

## Observed

`group=4`, 12 heads, head_dim 64, fp16, vs an fp32 CPU reference:

| extent | sticks | rows | max abs err | % elements >0.1 | |
|---:|---:|---:|---:|---:|---|
| 64  | 1 | 256  | 0.00724 | 0.00%  | ok |
| 128 | 2 | 512  | **1.75406** | **25.37%** | WRONG (intermittent) |
| 192 | 3 | 768  | 0.00678 | 0.00%  | ok |
| 256 | 4 | 1024 | 0.00715 | 0.00%  | ok |
| 320 | 5 | 1280 | **0.76863** | 15.82% | WRONG |
| 384 | 6 | 1536 | **0.65588** | 13.85% | WRONG |
| 448 | 7 | 1792 | **0.82254** | 12.04% | WRONG |
| 512 | 8 | 2048 | **0.68595** | 10.68% | WRONG |

`group=2, extent=512` also fails (0.62032), so it is not specific to `group=4`.
Attention outputs here are order-1, so these are wrong values, not rounding.

## `extent=128` is nondeterministic — please read before validating a fix

Three runs, each a fresh process with `/tmp/torchinductor_*` wiped, identical seed:

```
RESULT group=4 extent=128 max_abs_err=0.00961 ok
RESULT group=4 extent=128 max_abs_err=1.75406 WRONG
RESULT group=4 extent=128 max_abs_err=1.75406 WRONG
RESULT group=4 extent=512 max_abs_err=0.68595 WRONG   (3/3)
```

So a bad plan is selected on some compiles and not others — consistent with the tied
CP-SAT optima described in #4817. **A single passing run at `extent=128` does not
demonstrate a fix**; `extent=512` is the deterministic case to gate on.

## Error structure

At `group=4, extent=512`, the wrong elements are **alternate sequence rows, with the
parity flipping between batches** — exactly 256 of 512 positions in every batch:

```
batch 0: 256/512 seq positions wrong, first=1 last=511, non-contiguous
batch 1: 256/512 seq positions wrong, first=0 last=510, non-contiguous
batch 2: 256/512 seq positions wrong, first=1 last=511, non-contiguous
batch 3: 256/512 seq positions wrong, first=0 last=510, non-contiguous
```

All 12 heads and all 64 head_dim lanes are affected, so this looks like a row-indexing
or row-interleaving fault rather than a head- or lane-level one. The flat row dimension
is the dense flattening `extent * batch + sequence`, which the per-tile K restickify
reads directly (see the docstring on `test_grouped_sdpa_from_packed_rows` and the
padding work in #4685).

## Bisect

`ffe4d6ee` (parent of #4551) built from source, same script, same shapes:

| extent | `ffe4d6ee` (pre-#4551) | `main` |
|---:|---:|---:|
| 64  | 0.00724 ok | 0.00724 ok |
| 128 | **0.00961 ok** | **1.75406 WRONG** |
| 256 | 0.00715 ok | 0.00715 ok |
| 384 | **0.00483 ok** | **0.65588 WRONG** |
| 512 | **0.00626 ok** | **0.68595 WRONG** |

Pre-#4551 residuals are fp16 noise, i.e. correct rather than merely closer.

## Ruled out

- **Not the test reference.** The existing test compares against an fp16 CPU SDPA at
  `atol/rtol=0.1`. I checked that fp16 reference against fp32: they agree to **<=0.0005**
  at every shape above. The reference is sound.
- **Not fixed by #4812.** Identical failure on `17e157db`, `940c65ce` and `54fd5e12`
  (#4812), so it is not on the cost-model/planning axis that #4790/#4812 address.
- **Not caused by local changes.** Reproduced on pristine checkouts with no patches;
  verified by removing an unrelated local experiment entirely and re-running.

## Existing coverage understates this

`tests/inductor/test_building_blocks.py::TestBuildingBlocks::test_grouped_sdpa_from_packed_rows`
parametrises `(2,63), (2,64), (4,512)`. It catches `(4,512)` and never samples
`(4,128)`, which is wrong by more than twice as much. Extents 320/384/448 are also
uncovered. Suggest extending the parametrisation to at least 128 and one non-power-of-two
multi-stick extent.

Model-level tests do not catch it either: an encoder e2e suite (16 tests across
embedding, rerank and token-classification) passes on an affected build, because
embedding cosine >=0.99 and rerank score tolerances absorb a 0.686 absolute error on
individual attention outputs.

## Impact

These are production encoder extents. Any workload bucketing sequence length to
65-128 tokens (extent 128) or 257-512 tokens (extent 512) and packing a batch into one
row buffer is affected, and the failure is silent.

## Environment

- torch-spyre `17e157db` (also `940c65ce`, `54fd5e12`); clean at `ffe4d6ee`
- fp16, `SENCORES` default (32), 1 card, `LAYOUT_SOLVER` default (`cpsat`)

**Comment by gkumbhat** [comment]
## Correction: this is a deeptools defect, not a torch-spyre lowering bug

The bisect in the issue body is reproducible, but **my conclusion from it was wrong**.
The fix is in deeptools, at commit `779cd5a53c`. With that deeptools and no torch-spyre
change at all, every shape in the table above is clean:

| shape | earlier deeptools | deeptools `779cd5a53c` |
|---|---:|---:|
| `group=4, extent=512` | 0.68595 WRONG (3/3) | **0.00578 ok (3/3)** |
| `group=4, extent=128` | 1.75406 WRONG (3/3) | **0.00961 ok (3/3)** |

Confirmed against two independently produced builds carrying that fix, which give
byte-identical clean results.

**So #4551 did not introduce a defect** — it began generating code that reaches a broken
backend path. That is consistent with the bisect: `ffe4d6ee` was clean on the *same*
deeptools that `aabfe9cd` failed on, because the pre-#4551 lowering never emitted the
pattern that triggers it. It also explains why no cost-model change could fix it.

Please disregard the implication in the title and body that the fix belongs in
`decompositions.py`.

### The actionable constraint

**torch-spyre at or after #4551 requires deeptools >= `779cd5a53c` to be numerically
correct** at extents 128, 320, 384, 448 and 512. Pins predating #4551 are unaffected at
any extent.

### Still accurate in the body above

- The repro, the extent table and the error structure (alternate sequence rows wrong with
  the parity flipping between batches) all stand as observations on the older deeptools.
- `extent=128` is nondeterministic there (correct 1 run in 3). Applying #4817 pins the
  choice and makes it fail 3/3 — so #4817 is useful for reproducing this deterministically,
  though it does not fix it. #4812 does not fix it either. Both are cost-model-layer
  changes and the defect was in the backend.
- **The test-coverage gap is unrelated to deeptools and still worth closing:**
  `test_grouped_sdpa_from_packed_rows` parametrises `(2,63), (2,64), (4,512)` and never
  samples 128, 320, 384 or 448. On the older deeptools, `(4,128)` is wrong by 1.75 —
  more than twice the case the test does catch. A guard against a future recurrence would
  want at least 128 and one non-power-of-two multi-stick extent.
